### PR TITLE
Fix a bug for table statements in functions

### PR DESCRIPTION
**Pull request by alkino**
With the last patch we were not allocating memory and so it gives a bug.
Do it in a better way by putting the symbol inside the ast and running the symtab after

### CODE REVIEW

File: src/codegen/codegen_c_modif_visitor.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include "ast/function_block.hpp"
+#include "ast/string.hpp"
+#include "ast/table_statement.hpp"
+#include "codegen/codegen_c_modif_visitor.hpp"
+#include "visitors/visitor_utils.hpp"
+
+namespace nmodl {
+using namespace ast;
+
+void CodegenCModifVisitor::visit_function_block(FunctionBlock& node) {
```
Comment (pramodk) [inline src/codegen/codegen_c_modif_visitor.cpp:22]: `nitpick`: name could be `CodegenTransformVisitor` indicating all transformations/modifications related to code generation. And filename `codegen_transform_visitor.cpp/hpp`.

File: src/codegen/codegen_c_modif_visitor.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * \file
+ * \brief \copybrief nmodl::visitor::LocalToAssignedVisitor
+ */
+
+#include "visitors/ast_visitor.hpp"
+
+namespace nmodl {
+
+/**
+ * @addtogroup visitor_classes
```
Comment (pramodk) [inline src/codegen/codegen_c_modif_visitor.hpp:20]: is there a different for codegen?

File: src/codegen/codegen_c_modif_visitor.hpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * In the above example, `qt` is used as temporary variable to pass value from
+ * INITIAL block to PROCEDURE. This works fine in case of serial execution but
+ * in parallel execution we end up in race condition. To avoid this, we convert
+ * qt to ASSIGNED variable.
+ *
+ * \todo
+ *   - Variables like qt are often constant. As long as INITIAL block is executed
+ *     serially or qt is updated in atomic way then we don't have a problem.
+ */
+
```
Comment (pramodk) [inline src/codegen/codegen_c_modif_visitor.hpp:63]: needs to be updated

File: src/main.cpp [hunk 4]
```diff
@@ -556,6 +565,10 @@ int main(int argc, const char* argv[]) {
 
             else if (c_backend) {
                 logger->info("Running C backend code generator");
+                CodegenCModifVisitor{}.visit_program(*ast);
+                ast_to_nmodl(*ast, filepath("CModifVisitor"));
+                SymtabVisitor(update_symtab).visit_program(*ast);
+
                 CodegenCVisitor visitor(modfile,
```
Comment (pramodk) [inline src/main.cpp:572]: duplication?

File: src/codegen/CMakeLists.txt [hunk 5]
```diff
@@ -4,6 +4,7 @@
 add_library(
   codegen STATIC
   codegen_acc_visitor.cpp
+  codegen_c_modif_visitor.cpp
   codegen_c_visitor.cpp
   codegen_compatibility_visitor.cpp
   codegen_cuda_visitor.cpp
```
Comment (pramodk) [inline src/codegen/CMakeLists.txt:10]: a test would be great!

File: src/codegen/codegen_c_modif_visitor.hpp (hunk 3 above)
Comment (alkino) [reply src/codegen/codegen_c_modif_visitor.hpp:63]: Lol sorry

File: src/codegen/codegen_transform_visitor.hpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * @{
+ */
+
+/**
+ * \class CodegenTransformVisitor
+ * \brief Visitor to make last transformation to AST before codegen
+ *
+ * Modifications made:
+ * - add an argument to table if there are inside a function. In this case
+ *   the argument is the name of the function.
```
Comment (ohm314) [inline src/codegen/codegen_transform_visitor.hpp:30]:
```suggestion
 * - add an argument to the table if it is inside a function. In this case
 *   the argument is the name of the function.
```

File: src/codegen/CMakeLists.txt (hunk 5 above)
Comment (alkino) [reply src/codegen/CMakeLists.txt:10]: done

**Review by iomaganaris — APPROVED** [review]
LGTM 👍 
 This PR fixes #771

**Review by ohm314 — APPROVED** [review]
lgtm

**Pull request merged**